pysb/bng.py

```diff
@@ -91,14 +92,260 @@ class GenerateNetworkError(RuntimeError):
+
+        Parameters
+        ----------
+        overwrite: bool, optional
+            Overwrite existing network file, if any
+        """
+        try:
+            net_filename = self.base_filename + '.net'
+            self.action('generate_network', overwrite=overwrite)
+            with open(net_filename, 'r') as net_file:
+                bng_network = net_file.read()
+        finally:
+            if self.cleanup:
+                try:
+                    os.unlink(net_filename)
+                except OSError:
+                    # File may have been deleted already, or be open elsewhere
+                    pass
+        return bng_network
+
+    def action(self, action, **kwargs):
+        """
+        Generates a BNG action command and executes it through the console,
+        returning any console output
+
+        Parameters
+        ----------
+        action : string
+            The name of the BNG action function
+        kwargs : kwargs, optional
+            Arguments and values to supply to BNG
+        """
+        # Process BNG arguments into a string using supermethod
+        action_args = super(BngConsole, self).action(action, **kwargs)
+
+        # Execute the command via the console
+        self.console.sendline('action %s({%s})' % (action, action_args))
+
+        # Wait for the command to execute and return the result
+        return self._console_wait()
+
+
+class BngFileInterface(BngBaseInterface):
+    def __init__(self, model, verbose=False, output_dir=None,
+                 output_prefix=None, cleanup=True):
+        super(BngFileInterface, self).__init__(model, verbose, cleanup,
+                                               output_prefix)
+        self.command_queue = StringIO()
+        self.command_queue.write('begin actions\n')
+
+        # Create the BNGL file now so we know where to find it
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.bngl',
+                                         prefix=self.output_prefix,
+                                         dir=output_dir,
+                                         delete=False) as bng_file:
+            self.bng_filename = bng_file.name
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        In file mode, we execute all the queued commands when leaving the
+        context manager (i.e. now) by adding them to the BNGL
+        """
+        self.command_queue.write('end actions\ndone\n')
+        try:
+            # Generate BNGL file
+            with open(self.bng_filename, 'w') as bng_file:
+                bng_file.write(self.generator.get_content())
+                bng_file.write(self.command_queue.getvalue())
+
+            p = subprocess.Popen(['perl', _get_bng_path(), self.bng_filename],
+                                 cwd=os.path.dirname(self.bng_filename),
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE)
+            if self.verbose:
```

Comment:
I realize this is effectively the same code that was in `generate_network` before but I missed this issue when we reviewed the big dev branch merge.